src/transformStateWithClones.js

```diff
@@ -6,8 +6,39 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  const currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(currentState, action.extraData);
+        stateHistory.push({ ...currentState });
+        break;
+
+      case 'removeProperties':
+        for (const ch of action.keysToRemove) {
+          if (currentState.hasOwnProperty(ch)) {
+            delete currentState[ch];
+          }
+        }
+        stateHistory.push({ ...currentState });
+        break;
+
+      case 'clear':
+        for (const j in currentState) {
+          if (currentState.hasOwnProperty(j)) {
+            delete currentState[j];
+          }
+        }
+        stateHistory.push({ ...currentState });
+        break;
+    }
```

Comment:
Consider adding a default case in the switch statement for error handling. This will ensure that if an action type is not recognized, the function can handle it appropriately, such as by throwing an error or ignoring the action.